x-pack/legacy/plugins/lens/public/xy_visualization/to_expression.ts

```diff
@@ -122,82 +97,92 @@ export const buildExpression = (
-            xAxisOperation &&
-              xAxisOperation.isBucketed &&
-              xAxisOperation.scale &&
-              xAxisOperation.scale !== 'ordinal'
-          );
-
-          return {
-            type: 'expression',
-            chain: [
-              {
-                type: 'function',
-                function: 'lens_xy_layer',
-                arguments: {
-                  layerId: [layer.layerId],
-
-                  hide: [Boolean(layer.hide)],
-
-                  xAccessor: [layer.xAccessor],
-                  yScaleType: [
-                    getScaleType(metadata[layer.layerId][layer.accessors[0]], ScaleType.Ordinal),
-                  ],
-                  xScaleType: [
-                    getScaleType(metadata[layer.layerId][layer.xAccessor], ScaleType.Linear),
-                  ],
-                  isHistogram: [isHistogramDimension],
-                  splitAccessor: [layer.splitAccessor],
-                  seriesType: [layer.seriesType],
-                  accessors: layer.accessors,
-                  columnToLabel: [JSON.stringify(columnToLabel)],
+              ],
+            },
+          ],
+          layers: validLayers.map(layer => {
+            const columnToLabel: Record<string, string> = {};
+
+            if (frame) {
+              const datasource = frame.datasourceLayers[layer.layerId];
+              layer.accessors
+                .concat(layer.splitAccessor ? [layer.splitAccessor] : [])
+                .forEach(accessor => {
+                  const operation = datasource.getOperationForColumnId(accessor);
+                  if (operation && operation.label) {
+                    columnToLabel[accessor] = operation.label;
+                  }
+                });
+            }
+
+            const xAxisOperation =
+              frame &&
+              frame.datasourceLayers[layer.layerId].getOperationForColumnId(layer.xAccessor!);
+
+            const isHistogramDimension = Boolean(
+              xAxisOperation &&
+                xAxisOperation.isBucketed &&
+                xAxisOperation.scale &&
+                xAxisOperation.scale !== 'ordinal'
+            );
+
+            return {
+              type: 'expression',
+              chain: [
+                {
+                  type: 'function',
+                  function: 'lens_xy_layer',
+                  arguments: {
+                    layerId: [layer.layerId],
+
+                    hide: [Boolean(layer.hide)],
+
+                    xAccessor: [layer.xAccessor!],
+                    yScaleType: [
+                      getScaleType(metadata[layer.layerId][layer.accessors[0]], ScaleType.Ordinal),
+                    ],
+                    xScaleType: [
+                      getScaleType(metadata[layer.layerId][layer.xAccessor!], ScaleType.Linear),
+                    ],
+                    isHistogram: [isHistogramDimension],
+                    splitAccessor: [layer.splitAccessor!],
```

Comment:
I am not seeing why, this couldn't be `null`? Is there some implicit logic here, that I am missing, why `splitAccessor` can never be `null`?

Reply:
Fixed